### PR TITLE
New JSON API

### PR DESCRIPTION
**Pull request by markdroth**
Built on #21224.  Only the last commit is new.

Note: The json_reader_new.cc and json_writer_new.cc files are copies of json_reader.cc and json_writer.cc, so you may want to look at the diffs between them to see what's actually changed.

### CODE REVIEW

**Comment by markdroth** [comment]
Note: The json_reader_new.cc and json_writer_new.cc files are copies of json_reader.cc and json_writer.cc, so you may want to look at the diffs between them to see what's actually changed.

**Comment by markdroth** [comment]
Nico has been too busy to look at this. Esun, can you review instead? Thanks!

**Comment by veblush** [comment]
Thank you for the PR! LGTM!

File: src/core/lib/json/json_writer_new.cc [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+ * it to generate invalid json.
+ *
+ * Also, unlike the reader, the writer expects UTF-8 encoded input strings.
+ * These strings will be UTF-8 validated, and any invalid character will
+ * cut the conversion short, before any invalid UTF-8 sequence, thus forming
+ * a valid UTF-8 string overall.
+ */
+class JsonWriter {
+ public:
+  static UniquePtr<char> Dump(const Json& value, int indent);
```
[the remark this answers is not in the document]
Comment (markdroth) [reply src/core/lib/json/json_writer_new.cc]: Good idea.  Done.

**Comment by markdroth** [comment]
@nicolasnoble, Can you help me figure out why json_test_new.cc is failing on Windows?

https://source.cloud.google.com/results/invocations/27da2d15-b978-4af2-bc20-bb72ef1e6787/targets/%2F%2Ftest%2Fcore%2Fjson:json_test_new/tests

**Comment by markdroth** [comment]
Known issues: #21839 #21840

**Pull request merged**